src/crowdin_client.cpp

```diff
@@ -57,56 +65,71 @@
+#define OAUTH_SCOPE         "project+tm"
+#define OAUTH_CLIENT_ID     "k0uFz5HYQh0VzWgZmOpA"
+#define OAUTH_CLIENT_SECRET "p6Ko83cd4l4SD2TYeY8Fheffw8VuN3bML5jx7BeQ"
+//      any arbitrary unique unguessable string (e.g. UUID in hex)
+#define OAUTH_STATE         "948cf13ffffb47119d6cfa2b68898f67"
+//      The value of below macro should be set exactly as is (without quotes)
+//      to "Authorization Callback URL" of Crowdin application
+//      https://support.crowdin.com/enterprise/creating-oauth-app/
 #define OAUTH_URI_PREFIX    "poedit://auth/crowdin/"
-
-// Recursive extract files from /api/project/*/info response
-void ExtractFilesFromInfo(std::vector<std::wstring>& out, const json& r, const std::wstring& prefix)
-{
-    for (auto i : r["files"])
-    {
-        std::wstring name = prefix + str::to_wstring(i["name"]);
-        std::string node_type = i["node_type"];
-        if (node_type == "file")
-        {
-            out.push_back(name);
-        }
-        else if (node_type == "directory" || node_type == "branch")
-        {
-            ExtractFilesFromInfo(out, i, name + L"/");
-        }
-    }
-}
+#define OAUTH_AUTHORIZE_URL "/oauth/authorize" \
+	"?" "response_type" "=" "code" \
+	"&" "scope"         "=" OAUTH_SCOPE \
+	"&" "client_id"     "=" OAUTH_CLIENT_ID \
+	"&" "state"         "=" OAUTH_STATE \
+        "&" "redirect_uri"  "=" OAUTH_URI_PREFIX
 
 } // anonymous namespace
 
 
 std::string CrowdinClient::WrapLink(const std::string& page)
 {
-    std::string url("https://poedit.net/crowdin");
+    std::string url("https://accounts.crowdin.com");
     if (!page.empty() && page != "/")
-        url += "?u=" + http_client::url_encode(page);
+        url += page;
     return url;
 }
 
 
 class CrowdinClient::crowdin_http_client : public http_client
 {
 public:
-    crowdin_http_client(CrowdinClient& owner)
-        : http_client("https://api.crowdin.com"), m_owner(owner)
+    crowdin_http_client(CrowdinClient& owner, const std::string& url_prefix)
+        : http_client(url_prefix), m_owner(owner)
     {}
 
 protected:
     std::string parse_json_error(const json& response) const override
     {
-        std::string msg = response["error"]["message"];
-
-        // Translate commonly encountered messages:
-        if (msg == "Translations download is forbidden by project owner")
-            msg = _("Downloading translations is disabled in this project.").utf8_str();
+        cout << "\n\nJSON error: " << response << "\n\n";
+
+        auto it = response.find("error"),
+             end = response.end();
+
+        if(it == end) {
+            it = response.find("errors"); if(it == end) goto deflt;
+            end = it->end();
+            it = it->begin(); if(it == end) goto deflt;
+            end = it->end();
+            it = it->find("error"); if(it == end) goto deflt;
+            end = it->end();
+            it = it->find("errors"); if(it == end) goto deflt;
+            end = it->end();
+            it = it->begin();
```

Comment:
Well, because some API v2 error responses of even same request might need **sometime** e.g. `response["errors"][0]["error"]["errors"][0]["message"]` to get error message while **sometime** `response["error"]["message"]`. So when I just calling either of above to get message while it is really another one - it just crashes even if inside `try {} catch(...) {}`. i.e. I have not found any another reliable way to handle JSON schema mismatch to avoid crash with [nlohmann/json](https://github.com/nlohmann/json) lib. That is why I check here whether particular JSON key exists with `find()` or whether there are first element in JSON array with `begin()`. Or what else do you mean "explicitly checking for existing JSON keys" here? If something different than I mentioned - could you please provide some very simple example or so how you would handle this case in more straightforward manner.

Reply:
Marking as resolved it for a while just to easier track what else requires my attention yet. But please feel free to `Unresolve` it if you nevertheless see better solution here.

Reply:
I mean that this code is hard to understand - not only at glance but even when *paying attention*. An example of covered kinds of responses in a comment above it would help, but self-explanatory code would be better.

The JSON library's readme says this about about access:

> use function at() to access the object values rather than operator[]. In case a key does not exist, at throws an exception that you can handle, whereas operator[] exhibits undefined behavior.

In other words, my use of `operator[]` all over the place is wrong, and `at()` is safe to use. Alternatively, but that doesn't apply to this specific case, `value()` can be used if you know what the default value in absence of the key should be.

Reply:
`at()` worked like charm! thanks a lot for so great hint! I don't know how to missed it in the [nlohmann/json README](https://github.com/nlohmann/json) (might be because just googled and looked in [nlohmann/json reference](https://nlohmann.github.io/json), though another hint for me to always read README more carefully :) ). So it's also so great elegant solution for all JSON schema validation issues around whole [crowdin_client.cpp](https://github.com/tetatetit/poedit/blob/crowdin-new-oauth-api-xliff/src/crowdin_client.cpp) (i.e. when server response has different JSON structure/schema than expected).

Force-pushed to commit it originally introduced 8f42ca0ea6603fdbf490efc98a69a1e7be5c9278